### PR TITLE
dbus activation: add dbus service file

### PR DESCRIPTION
**Pull request by pesader**
First, refactor the `DesktopEntry` class into a more generic `InstallableResource` class, that doesn't have pre-defined values for `source_file` and `target_file`, and can be used for both the desktop entry file and the dbus service file.

Then, create an instance of this new class for the dbus file and install it to a subdirectory of the user's runtime directory $XDG_RUNTIME_DIR, where the dbus daemon looks for these files.

If merged, this would solve issue #304.

### CODE REVIEW

**Comment by amezin** [comment]
Thank you for your effort.

Looks good on the first sight, but I need to test it manually a bit.

File: ddterm/org.gnome.Shell.Extensions.ddterm.service [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[D-BUS Service]
+Name=org.gnome.Shell.Extensions.ddterm
```
Comment (amezin) [inline ddterm/org.gnome.Shell.Extensions.ddterm.service:2]: There's currently nothing with well-known name `org.gnome.Shell.Extensions.ddterm`. The app registers on the bus under the name `com.github.amezin.ddterm` https://github.com/ddterm/gnome-shell-extension-ddterm/blob/master/ddterm/app/application.js#L310

File: ddterm/shell/extension.js [hunk 2]
```diff
@@ -245,8 +240,28 @@ function enable() {
...
+        Me.dir.get_child('ddterm').get_child('org.gnome.Shell.Extensions.ddterm.service'),
+        Gio.File.new_for_path(GLib.build_filenamev(
+            [
+                GLib.get_user_runtime_dir(),
+                'dbus-1',
+                'services',
+                'org.gnome.Shell.Extensions.ddterm.service',
+            ]))
+    );
+    dbus_service.install();
```
Comment (amezin) [inline ddterm/shell/extension.js:264]: d-bus daemon docs for `$XDG_RUNTIME_DIR/dbus-1/services`:

> Also unlike the other standard session service directories, this directory is never monitored with inotify(7) or similar APIs. Programs that create service files in this directory while a dbus-daemon is running are expected to call the dbus-daemon's ReloadConfig() method after they have made changes.

https://dbus.freedesktop.org/doc/dbus-daemon.1.html

**Review by amezin — CHANGES_REQUESTED** [review]
Try testing D-Bus activation using d-feet. When ddterm app isn't running (i. e. all tabs closed), d-feet should show the service on the session bus as activatable. And selecting it should start the application.

Also, when D-Bus activation starts working, it makes sense to enable it in the `.desktop` file.

Would `Exec=` in the `.desktop` file still be necessary?

**Comment by pesader** [comment]
> Try testing D-Bus activation using d-feet. When ddterm app isn't running (i. e. all tabs closed), d-feet should show the service on the session bus as activatable. And selecting it should start the application.

Looks like it is working :D

https://user-images.githubusercontent.com/65264536/217408480-35611841-4725-488e-b48f-f3127710e2fb.mp4

> Also, when D-Bus activation starts working, it makes sense to enable it in the .desktop file.

Done in 5a7ea73.

> Would Exec= in the .desktop file still be necessary?

From what I read in the [Freedesktop Spec](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html), the Exec line must be kept to maintain compatibility with implementations that don't handle the DBusActivatable line properly. To quote them:

> If the value is true then implementations should ignore the Exec key and send a D-Bus message to launch the application. See [D-Bus Activation](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#dbus) for more information on how this works. Applications should still include Exec= lines in their desktop files for compatibility with implementations that do not understand the DBusActivatable key.

File: ddterm/shell/extension.js (hunk 2 above)
Comment (pesader) [reply ddterm/shell/extension.js:264]: Thanks, I did not know that! Done in 2f0acb9.

File: ddterm/org.gnome.Shell.Extensions.ddterm.service (hunk 1 above)
Comment (pesader) [reply ddterm/org.gnome.Shell.Extensions.ddterm.service]: In e0c2db5, I changed the well-known name (and the filename to match it).

**Comment by amezin** [comment]
> Applications should still include Exec= lines in their desktop files for compatibility with implementations that do not understand the DBusActivatable key.

Hm. But the only "implementation" we need to care about is GNOME Shell.

**Comment by pesader** [comment]
> Hm. But the only "implementation" we need to care about is GNOME Shell.

Yeah, makes sense. Your call then :)

**Pull request merged**